### PR TITLE
[Flink-2971][table] Add outer joins to the Table API

### PR DESCRIPTION
**Pull request by dawidwys**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [ ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by fhueske** [comment]
Thanks for this PR, @dawidwys! I haven't had a look yet but will do soon. 

Just a quick comment: this PR touches some files that PR #1958 is also modifying (mostly on the API level, not the runtime and optimization code). PR #1958 is a bigger change but in good shape and should be mergable soon. I would like to merge #1958 before this one, so you will need to rebase once this has happened. That should not be too much work, just wanted to let you know in advance.

Thanks, Fabian

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetJoin.scala [hunk 1]
```diff
@@ -159,9 +158,20 @@ class DataSetJoin(
...
 
+    val (joinOperator, nullCheck) = joinType match {
+      case JoinRelType.INNER => (leftDataSet.join(rightDataSet), false)
+      case JoinRelType.LEFT => (leftDataSet.leftOuterJoin(rightDataSet), true)
+      case JoinRelType.RIGHT => (leftDataSet.rightOuterJoin(rightDataSet), true)
+      case JoinRelType.FULL => (leftDataSet.fullOuterJoin(rightDataSet), true)
+    }
+
+    if (nullCheck && !config.getNullCheck) {
+      throw new TableException("Null check in TableEnvironment must be enabled for outer joins.")
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetJoin.scala]: TableEnvironment should be TableConfig

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 2]
```diff
@@ -334,6 +335,156 @@ class Table(
...
+  def join(right: Table, joinPredicate: Expression): Table = {
+    join(right, joinPredicate, JoinType.INNER)
+  }
+
+  /**
+    * Joins two [[Table]]s. Similar to an SQL left outer join. The fields of the two joined
+    * operations must not overlap, use [[as]] to rename fields if necessary.
+    *
+    * Note: Both tables must be bound to the same [[TableEnvironment]] and its [[TableConfig]] must
+    * have nullCheck enables.
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: should be "... must have nullCheck enableD"

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 3]
```diff
@@ -334,6 +335,156 @@ class Table(
...
+    *
+    * {{{
+    *   left.fullOuterJoin(right, 'a === 'b && 'c > 3).select('a, 'b, 'd)
+    * }}}
+    */
+  def fullOuterJoin(right: Table, joinPredicate: Expression): Table = {
+    join(right, joinPredicate, JoinType.FULL_OUTER)
+  }
+
+  private def flinkJoinTypeToCalcite(joinType: JoinType) = joinType match {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: I think you can also directly use Calcite's `JoinRelType` instead of Flink's `JoinType`.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/JoinITCase.scala [hunk 4]
```diff
@@ -259,14 +260,21 @@ class JoinITCase(
...
 
-  @Test(expected = classOf[PlanGenException])
+  @Test
   def testLeftOuterJoin(): Unit = {
 
     val env = ExecutionEnvironment.getExecutionEnvironment
     val tEnv = TableEnvironment.getTableEnvironment(env, config)
+    tEnv.getConfig.setNullCheck(true)
 
     val sqlQuery = "SELECT c, g FROM Table3 LEFT OUTER JOIN Table5 ON b = e"
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/sql/JoinITCase.scala]: Switch Table3 and Table5 to have also null values in the result

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala [hunk 5]
```diff
@@ -106,9 +107,9 @@ class JoinITCase(mode: TestExecutionMode) extends MultipleProgramsTestBase(mode)
     val ds2 = CollectionDataSets.get5TupleDataSet(env).toTable(tEnv, 'd, 'e, 'f, 'g, 'h)
 
     ds1.join(ds2)
-      // must fail. Field 'foo does not exist
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala]: I think this indention was intended. Can you undo the change?

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala [hunk 6]
```diff
@@ -271,4 +270,86 @@ class JoinITCase(mode: TestExecutionMode) extends MultipleProgramsTestBase(mode)
...
+  @Test
+  def testLeftJoinWithFilterInJoinCondition(): Unit = {
+    val env: ExecutionEnvironment = ExecutionEnvironment.getExecutionEnvironment
+    val tEnv = TableEnvironment.getTableEnvironment(env)
+    tEnv.getConfig.setNullCheck(true)
+
+    val ds1 = CollectionDataSets.get3TupleDataSet(env).toTable(tEnv, 'a, 'b, 'c)
+    val ds2 = CollectionDataSets.get5TupleDataSet(env).toTable(tEnv, 'd, 'e, 'f, 'g, 'h)
+
+    val joinT = ds1.leftOuterJoin(ds2, 'a < 3 && 'b === 'd.cast(TypeInformation.of(classOf[Long])))
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala]: Can you change the query such that the result has a null field?

**Comment by fhueske** [comment]
Hi @dawidwys, 
the runtime code, rule, API methods, and the tests look very good. :-)

PR #1958 should be ready to be merged (waiting for @twalthr to give his OK). I think you can rebase your code on top of #1958. 

One last comment with respect to style fixes. Unfortunately, we do not have a strict code style in place that is automatically enforced and contributors follow sometimes different styles. We try to keep style changes in PRs to a minimum. Some changes make absolutely sense, but other changes might be reverted by the next person going over the file. Also style changes can distract from important changes. There is no hard rule for what to change, but a good rule of thumb is to leave code as it is if in doubt. 

Thanks, Fabian

**Comment by fhueske** [comment]
Hi @dawidwys, do you have some time to work on this PR this week?
I will be away next week + a couple of days and would really like to merge it before.
Thanks, Fabian

**Comment by dawidwys** [comment]
Hi @fhueske. Yes I will update this PR today. Sorry I haven't done it earlier but I was away for the past week.

**Comment by fhueske** [comment]
No worries and thanks @dawidwys!

**Comment by dawidwys** [comment]
@fhueske I have uploaded the updated PR. Unfortunately there are some strange VM crashes, that I think are not related to the changes.

**Comment by fhueske** [comment]
Yes, the failures are a known issue at the moment :-/ Hope this will be fixed soon.
I'll run the tests later myself to check that the build passes.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/fieldExpression.scala [hunk 7]
```diff
@@ -17,9 +17,9 @@
  */
 package org.apache.flink.api.table.expressions
 
-import org.apache.calcite.rex.RexNode
+import org.apache.calcite.rel.RelNode
 import org.apache.calcite.tools.RelBuilder
-
+import org.apache.calcite.rex.{RexInputRef, RexNode}
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/expressions/fieldExpression.scala]: No need to change the imports in this file.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 8]
```diff
@@ -269,22 +269,60 @@ case class Join(
...
-    joinType match {
-      case JoinType.INNER => left.output ++ right.output
-      case j => throw new ValidationException(s"Unsupported JoinType: $j")
+    left.output ++ right.output
+  }
+
+  private case class JoinFieldReference(
+    name: String,
+    resultType: TypeInformation[_],
+    left: RelNode,
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I think JoinFieldReference should rather use `LogicalNode` instead of `RelNode` to refer to the inputs of the join.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 9]
```diff
@@ -269,22 +269,60 @@ case class Join(
...
-      case JoinType.INNER =>
-        left.construct(relBuilder)
-        right.construct(relBuilder)
-        relBuilder.join(JoinRelType.INNER,
-          condition.map(_.toRexNode(relBuilder)).getOrElse(relBuilder.literal(true)))
-      case _ =>
-        throw new ValidationException(s"Unsupported JoinType: $joinType")
+    left.construct(relBuilder)
+    right.construct(relBuilder)
+    val partialFunction: PartialFunction[Expression, Expression] = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I think the resolution into `JoinFieldReference` should happen in the `resolveExpression()` method / phase and not in `construct()`. See the docs in `LogicalNode`. What do you think @yjshen?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 10]
```diff
@@ -269,22 +269,60 @@ case class Join(
...
+        relBuilder.peek(2, 0),
+        relBuilder.peek(2, 1))
     }
+
+    val transformedExpression = condition.map(_.postOrderTransform(partialFunction))
+                                .map(_.toRexNode(relBuilder)).getOrElse(relBuilder.literal(true))
+
+    relBuilder.join(flinkJoinTypeToCalcite(joinType), transformedExpression)
+  }
+
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: can you implement the `validate()` method and check if the join condition (if set) consists only of join predicates and at least one equality join predicate? Otherwise we would fail later during optimization.

So something like: `(left.a == right.b AND left.c < right.d)` would be OK but the following would fail:
- `(left.a < right.b)` // no equi join
- `(left.a == right.b or left.c == right.d)` // disjunctive preds are not supported atm
- `(left.a == right.b and left.a > 10)` // non-join preds would be OK, but I think it would be cleaner to do that in `where()`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetJoin.scala [hunk 11]
```diff
@@ -77,7 +76,7 @@ class DataSetJoin(
   }
 
   override def toString: String = {
-    s"Join(where: ($joinConditionToString), join: ($joinSelectionToString))"
+    s"$joinName(where: ($joinConditionToString), join: ($joinSelectionToString))"
   }
 
   override def explainTerms(pw: RelWriter): RelWriter = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetJoin.scala]: can you add the `joinType` to `explainTerms()` ?

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetJoinRule.scala [hunk 12]
```diff
@@ -61,21 +51,22 @@ class DataSetJoinRule
...
-      convRight,
-      rel.getRowType,
-      join.getCondition,
-      join.getRowType,
-      joinInfo,
-      joinInfo.pairs.toList,
-      JoinType.INNER,
-      null,
-      description)
+        new DataSetJoin(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/rules/dataSet/DataSetJoinRule.scala]: the indention is off

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala [hunk 13]
```diff
@@ -251,12 +249,157 @@ class Table(
...
+  /**
+    * Joins two [[Table]]s. Similar to an SQL join. The fields of the two joined
+    * operations must not overlap, use [[as]] to rename fields if necessary.
+    *
+    * Note: Both tables must be bound to the same [[TableEnvironment]].
+    *
+    * Example:
+    *
+    * {{{
+    *   left.join(right, "a = b && c > 3")
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/table.scala]: If we decide to only allow join predicates as join condition, we should remove the `c > 3` predicate.

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala [hunk 14]
```diff
@@ -191,12 +192,12 @@ class JoinITCase(mode: TestExecutionMode) extends MultipleProgramsTestBase(mode)
     val ds2 = CollectionDataSets.get5TupleDataSet(env).toTable(tEnv, 'd, 'e, 'f, 'g, 'h)
 
     val joinT = ds1.join(ds2)
-      .where('a === 'd)
-      .groupBy('a, 'd)
-      .select('b.sum, 'g.count)
+                .where('a === 'd)
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala]: please revert this indention change

File: flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala [hunk 15]
```diff
@@ -271,4 +270,89 @@ class JoinITCase(mode: TestExecutionMode) extends MultipleProgramsTestBase(mode)
...
+      "Comment#3,null\n" + "Comment#4,null\n" + "Comment#5,null\n" + "Comment#6,null\n" +
+      "Comment#7,null\n" + "Comment#8,null\n" + "Comment#9,null\n" + "Comment#10,null\n" +
+      "Comment#11,null\n" + "Comment#12,null\n" + "Comment#13,null\n" + "Comment#14,null\n" +
+      "Comment#15,null\n"
+    val results = joinT.toDataSet[Row].collect()
+    TestBaseUtils.compareResultAsText(results.asJava, expected)
+  }
+
+  @Test
+  def testLeftJoinWithFilterInJoinCondition(): Unit = {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/test/scala/org/apache/flink/api/scala/batch/table/JoinITCase.scala]: This test can be removed if we do not allow non-join predicates in join conditions.

**Comment by fhueske** [comment]
Thanks for the update @dawidwys!
The resolving / validation phase needs a bit of rewrite / extension. Other than that it looks good. The other comments are just minor issues.

Thanks, Fabian

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 16]
```diff
@@ -269,22 +269,57 @@ case class Join(
...
+    override def toString = s"'$name"
+
+    override def toRexNode(implicit relBuilder: RelBuilder): RexNode = {
+      val joinInputField = left.output.zipWithIndex.find(_._1.name == name)
+                           .orElse(
+                             right.output.zipWithIndex.find(_._1.name == name)
+                             .map(x => (x._1, left.output.length + x._2))).get
+
+      new RexInputRef(joinInputField._2,
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: `Expression.toRexNode()` is called in `LogicalNode.construct()`. That means, that the input operators are on the top of the `RelBuilder` operations stack. So we can use, `RelBuilder.field(int inputCount, int inputOrdinal, String fieldName)` and do not have to deal with the offset computation ourselves. 

I would also do the look up of the input-local index in the constructor of `JoinFieldReference` and add a flag that indicates whether the field is from the left or right input. That will be helpful when validating for join conditions.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 17]
```diff
@@ -298,11 +333,34 @@ case class Join(
...
+    } else if (ambiguousName.nonEmpty) {
       failValidation(s"join relations with ambiguous names: ${ambiguousName.mkString(", ")}")
     }
+
+    resolvedJoin.condition.foreach(testJoinCondition(_))
     resolvedJoin
   }
+
+  private def testJoinCondition(expression: Expression): Unit = {
+    def checkIfJoinCondition(exp : Expression) = if (exp.children.exists(!_.isInstanceOf[JoinFieldReference])) {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: The condition here should be: all (of the two) children are `JoinFieldReference` and one is on left and the other on the right input (that's where the input flag in `JoinFieldReference` is helpful).

I would also change the type of `exp` to `BinaryComparison`

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 18]
```diff
@@ -298,11 +333,34 @@ case class Join(
...
+      failValidation(s"Only join predicates supported. For non-join predicates use Table#where.")
+    }
+
+    var equiJoinFound = false
+    def validateConditions(exp: Expression) : Unit = exp match {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Checking whether it is an equi-join is a bit more difficult. For instance `l.a = r.b AND (l.c = r.d OR l.c = r.e)` is a valid equi-join condition. The first condition `(l.a = r.b)` can be pushed in the DataSet join (`left.join(right).where(a).equalTo(b)`) and the remaining condition can be evaluated in the `JoinFunction`.

To do the check, the predicate should be first normalized into "Conjunctive Normal Form" (CNF). Once that is done, the check is easy.

**Comment by fhueske** [comment]
Hi @dawidwys! Thanks for the fast update! Looks mostly good. I suggested one simplification with the `JoinFieldReference` and the validation needs to be improved a bit. Let me know if you have questions.

Thanks, Fabian

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala (hunk 16 above)
Comment (dawidwys) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Regarding the `RelBuilder.field` I also thought it is like you said, but unfortunately in case of Join it is not. The `RelBuilder.field` returns index to seperate inputs, while in `Join` operation it should point to  input being concatenation of both inputs. That is why I compute the index myself.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala (hunk 18 above)
Comment (dawidwys) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I am not sure if the normalization to CNF is necessary. Calcite does not do it, but searches just the `AND` operators subtrees for equi conditions. E.g. For condition (l.a = r.b AND l.c =r.d) OR (l.a = r.b AND l.e = r.f) after transformation to CNF it will be possible to find equi condition, but Calcite is not finding it (see `RelOptUtil#splitJoinCondition`) and results in failing DataSet join.

My proposition would be:

1. traverse whole tree(both `OR` and `AND` branches) checking if only join-conditions exists
2. check if equi-condition exists in `AND` branch

What do you think @fhueske ?

**Comment by dawidwys** [comment]
Thanks @fhueske for your comments. I uploaded another version with my resolutions to issues you've raised. Still happy to discuss.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 19]
```diff
@@ -272,23 +272,46 @@ case class Join(
...
+                           .getOrElse(
+                             throw new NoSuchElementException(s"""Could not find field: $name"""))
+                           .asInstanceOf[Int]
+
+      new JoinFieldReference(name, resultType, left, right, joinInputField)
+    }
+
+  }
+
   private case class JoinFieldReference(
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I am sorry, I think my last comment was confusing. Actually, I misinterpreted the semantics of the `RelBuilder.field(int, int, String)` method. I thought it would automatically add the offset of the left input. But apparently it doesn't...

How about we change the `JoinFieldReference` to this:

```
private case class JoinFieldReference(
      name: String,
      resultType: TypeInformation[_],
      left: LogicalNode,
      right: LogicalNode) extends Attribute {

    val isFromLeftInput = left.output.map(_.name).contains(name)

    val (indexInInput, indexInJoin) = if (isFromLeftInput) {
        val indexInLeft = left.output.map(_.name).indexOf(name)
        (indexInLeft, indexInLeft)
      } else {
        val indexInRight = right.output.map(_.name).indexOf(name)
        (indexInRight, indexInRight + left.output.length)
      }

    override def toString = s"'$name"

    override def toRexNode(implicit relBuilder: RelBuilder): RexNode = {

      // look up type of field
      val fieldType = relBuilder.field(2, if (isFromLeftInput) 0 else 1, name).getType()
      // create a new RexInputRef with index offset
      new RexInputRef(indexInJoin, fieldType)
    }

    override def withName(newName: String): Attribute = {
      if (newName == name) {
        this
      } else {
        JoinFieldReference(newName, resultType, left, right)
      }
    }
  }
```

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 20]
```diff
@@ -342,21 +368,29 @@ case class Join(
   }
 
   private def testJoinCondition(expression: Expression): Unit = {
-    def checkIfJoinCondition(exp : Expression) = if (exp.children.exists(!_.isInstanceOf[JoinFieldReference])) {
-      failValidation(s"Only join predicates supported. For non-join predicates use Table#where.")
-    }
+    def checkIfJoinCondition(exp : BinaryComparison) =
+      if (exp.children match {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: The condition looks good but can be a bit simplified like this:
```
exp.children match {
  case (x: JoinFieldReference) :: (y: JoinFieldReference) :: Nil
    if x.isFromLeftInput != y.isFromLeftInput
      => // predicate references both inputs. All good!
  case _ => failValidation(
    s"Only join predicates supported. For non-join predicates use Table#where.")
}
```

**Comment by dawidwys** [comment]
I applied your changes @fhueske . If you have any comments regarding the equi-join validation I will address/apply them in the evening.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 21]
```diff
@@ -342,21 +368,29 @@ case class Join(
...
+          x.belongsToLeft == y.belongsToLeft
+        case _ => true
+      }) {
+        failValidation(s"Only join predicates supported. For non-join predicates use Table#where.")
+      }
 
     var equiJoinFound = false
-    def validateConditions(exp: Expression) : Unit = exp match {
-      case x: And => x.children.foreach(validateConditions(_))
+    def validateConditions(exp: Expression, isAndBranch: Boolean): Unit = exp match {
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: I'm sorry but this check does not catch all cases. For example, it would let `l.a = r.b OR (l.c = r.d && l.e = r.f)` pass.

**Comment by fhueske** [comment]
Hi @dawidwys, the equi join check is a bit more involved. I implemented a CNF converter based on the algorithm described here: http://cs.jhu.edu/~jason/tutorials/convert-to-CNF (see my [branch](https://github.com/fhueske/flink/tree/tableOuter) ).
You can try to do implement the algorithm as well. Alternatively, we can also use my version.

Another thing we need to fix is the documentation, i.e., extending the Table API description and removing the outer join restriction in the SQL section in `docs/apis/table.md`.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala (hunk 21 above)
Comment (dawidwys) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: You're absolutely right, my mistake, sorry! I fixed it.

**Comment by dawidwys** [comment]
I am not sure if you've seen my comment(due it was to some previous commit version). Let me paste my doubts about CNF one more time:

> I am not sure if the normalization to CNF is necessary. Calcite does not do it, but searches just the AND operators subtrees for equi conditions. E.g. For condition (l.a = r.b AND l.c =r.d) OR (l.a = r.b AND l.e = r.f) after transformation to CNF it will be possible to find equi condition, but Calcite is not finding it (see RelOptUtil#splitJoinCondition) and results in failing DataSet join.
> 
> My proposition would be:
> 
> traverse whole tree(both OR and AND branches) checking if only join-conditions exists
> check if equi-condition exists in AND branch

**Comment by fhueske** [comment]
Ah, sorry. No, I didn't see your comment. Will think about that and reply later. Thanks! :-)

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 22]
```diff
@@ -298,11 +338,40 @@ case class Join(
...
+    resolvedJoin.condition.foreach(testJoinCondition(_))
     resolvedJoin
   }
+
+  private def testJoinCondition(expression: Expression): Unit = {
+    def checkIfJoinCondition(exp : BinaryComparison) = exp.children match {
+        case (x : JoinFieldReference) :: (y : JoinFieldReference) :: Nil
+          if x.isFromLeftInput != y.isFromLeftInput => Unit
+        case _ => failValidation(
+          s"Only join predicates supported. For non-join predicates use Table#where.")
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Add the condition that causes the failure to the error message.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala [hunk 23]
```diff
@@ -298,11 +338,40 @@ case class Join(
...
+          equiJoinFound = true
+        }
+        checkIfJoinCondition(x)
+      case x: BinaryComparison => checkIfJoinCondition(x)
+      case x => failValidation(s"Unsupported condition type: ${x.getClass.getSimpleName}.")
+    }
+
+    validateConditions(expression, isAndBranch = true)
+    if (!equiJoinFound) {
+      failValidation(s"At least one equi-join required.")
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala]: Add the whole join predicate to the error message?

**Comment by fhueske** [comment]
Thanks for looking into how Calcite splits the join conditions. You are right, we do not need to be "smarter" than Calcite and approve predicates that will be later rejected. So +1 for your suggestion.

Apart from the documentation update I would only extend the error messages with the conditions that cause the failure. After that the PR should be good to merge.

File: docs/apis/table.md [hunk 24]
```diff
@@ -423,14 +423,50 @@ Table result = in.groupBy("a").select("a, b.sum as d");
...
 Table right = tableEnv.fromDataSet(ds2, "d, e, f");
 Table result = left.join(right).where("a = d").select("a, b, e");
 {% endhighlight %}
       </td>
     </tr>
+    
+    <tr>
+      <td><strong>RightOuterJoin</strong></td>
+      <td>
+        <p>Similar to a SQL FULL OUTER JOIN clause. Joins two tables. Both tables must have distinct field names and at least one equality join predicate must be defined.</p>
```
Comment (fhueske) [inline docs/apis/table.md]: `FULL OUTER JOIN` should be `RIGHT OUTER JOIN`

File: docs/apis/table.md [hunk 25]
```diff
@@ -423,14 +423,50 @@ Table result = in.groupBy("a").select("a, b.sum as d");
...
+Table right = tableEnv.fromDataSet(ds2, "d, e, f");
+Table result = left.rightOuterJoin(right, "a = d").select("a, b, e");
+{% endhighlight %}
+      </td>
+    </tr>
+    
+    <tr>
+      <td><strong>LeftOuterJoin</strong></td>
+      <td>
+        <p>Similar to a SQL FULL OUTER JOIN clause. Joins two tables. Both tables must have distinct field names and at least one equality join predicate must be defined.</p>
```
Comment (fhueske) [inline docs/apis/table.md]: `FULL OUTER JOIN` should be `LEFT OUTER JOIN`

File: docs/apis/table.md [hunk 26]
```diff
@@ -542,6 +578,42 @@ val result = left.join(right).where('a === 'd).select('a, 'b, 'e);
 {% endhighlight %}
       </td>
     </tr>
+    
+    <tr>
+      <td><strong>RightOuterJoin</strong></td>
+      <td>
+        <p>Similar to a SQL FULL OUTER JOIN clause. Joins two tables. Both tables must have distinct field names and at least one equality join predicate must be defined.</p>
```
Comment (fhueske) [inline docs/apis/table.md]: `FULL OUTER JOIN` should be `RIGHT OUTER JOIN`

File: docs/apis/table.md [hunk 27]
```diff
@@ -542,6 +578,42 @@ val result = left.join(right).where('a === 'd).select('a, 'b, 'e);
...
+{% highlight scala %}
+val left = tableEnv.fromDataSet(ds1, 'a, 'b, 'c)
+val right = tableEnv.fromDataSet(ds2, 'd, 'e, 'f)
+val result = left.rightOuterJoin(right, 'a = 'd).select('a, 'b, 'e)
```
Comment (fhueske) [inline docs/apis/table.md]: `=` should be `===`

File: docs/apis/table.md [hunk 28]
```diff
@@ -542,6 +578,42 @@ val result = left.join(right).where('a === 'd).select('a, 'b, 'e);
...
+{% endhighlight %}
+      </td>
+    </tr>
+    
+    <tr>
+      <td><strong>LeftOuterJoin</strong></td>
+      <td>
+        <p>Similar to a SQL FULL OUTER JOIN clause. Joins two tables. Both tables must have distinct field names and at least one equality join predicate must be defined.</p>
```
Comment (fhueske) [inline docs/apis/table.md]: `FULL OUTER JOIN` should be `LEFT OUTER JOIN`

File: docs/apis/table.md [hunk 29]
```diff
@@ -542,6 +578,42 @@ val result = left.join(right).where('a === 'd).select('a, 'b, 'e);
...
+{% highlight scala %}
+val left = tableEnv.fromDataSet(ds1, 'a, 'b, 'c)
+val right = tableEnv.fromDataSet(ds2, 'd, 'e, 'f)
+val result = left.leftOuterJoin(right, 'a = 'd).select('a, 'b, 'e)
```
Comment (fhueske) [inline docs/apis/table.md]: `=` -> `===`

File: docs/apis/table.md [hunk 30]
```diff
@@ -542,6 +578,42 @@ val result = left.join(right).where('a === 'd).select('a, 'b, 'e);
...
+    </tr>
+
+    <tr>
+      <td><strong>FullOuterJoin</strong></td>
+      <td>
+        <p>Similar to a SQL FULL OUTER JOIN clause. Joins two tables. Both tables must have distinct field names and at least one equality join predicate must be defined.</p>
+{% highlight scala %}
+val left = tableEnv.fromDataSet(ds1, 'a, 'b, 'c)
+val right = tableEnv.fromDataSet(ds2, 'd, 'e, 'f)
+val result = left.fullOuterJoin(right, 'a = 'd).select('a, 'b, 'e)
```
Comment (fhueske) [inline docs/apis/table.md]: `=` -> `===`

**Comment by fhueske** [comment]
Thanks for the update @dawidwys. I found a few minor issues in the docs. Will fix them and merge this PR.

**Comment by dawidwys** [comment]
Really sorry for those mistakes :( Thanks for fixing them.
